### PR TITLE
[Notifier] Add Bandwidth bridge

### PR DESCRIPTION
**Pull request by gnito-org**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |  
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/17483
| Recipe         | https://github.com/symfony/recipes/pull/1153

This PR add the notifier bridge for [Bandwidth](https://www.bandwidth.com/) (SMS).

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        EventDispatcherInterface $dispatcher = null
+    )
+    {
+        $this->username = $username;
+        $this->password = $password;
+        $this->from = $from;
+        $this->accountId = $accountId;
+        $this->applicationId = $applicationId;
+        $this->priority = $priority;
+        parent::__construct(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: One line please

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    private string $accountId;
+
+    private string $applicationId;
+
+    private ?string $priority;
+
+    private string $password;
+
+    public function __construct(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php:32]: Please use constructor property promotion whenever possible

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        $this->priority = $priority;
+        parent::__construct(
+            $client,
+            $dispatcher
+        );
+    }
+
+    public function __toString(): string
+    {
+        return sprintf(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: One line please

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            'bandwidth://%s?from=%s&account_id=%s&application_id=%s&priority=%s',
+            $this->getEndpoint(),
+            $this->from,
+            $this->accountId,
+            $this->applicationId,
+            $this->priority
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: What if priority is null?

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        return $message instanceof SmsMessage;
+    }
+
+    /**
+     * https://dev.bandwidth.com/apis/messaging/#tag/Messages/operation/createMessage
+     */
+    protected function doSend(MessageInterface $message): SentMessage
+    {
+        if (!$message instanceof SmsMessage) {
+            throw new UnsupportedMessageTypeException(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: One line please

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        $options['applicationId'] = $options['application_id'] ?? $this->applicationId;
+        unset($options['application_id']);
+        if (!isset($options['priority']) && $this->priority) {
+            $options['priority'] = $this->priority;
+        }
+        if (!preg_match(
+            '/^\+[1-9]\d{1,14}$/',
+            $this->from
+        )) {
+            throw new InvalidArgumentException(
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: One line, same for many other spots, we can more easily avoid merge conflicts

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        string $from = 'from'
+    ): BandwidthTransport
+    {
+        return new BandwidthTransport(
+            username     : 'username',
+            password     : 'password',
+            from         : $from,
+            accountId    : 'account_id',
+            applicationId: 'application_id',
+            priority     : 'priority',
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php]: No need to use named arguments here

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/composer.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
+{
+    "name": "symfony/bandwidth-notifier",
+    "type": "symfony-notifier-bridge",
+    "description": "Symfony Bandwidth Notifier Bridge",
+    "keywords": [
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/composer.json:5]: One line

File: src/Symfony/Component/Notifier/LICENSE [hunk 9]
```diff
@@ -1,4 +1,4 @@
-Copyright (c) 2019-2022 Fabien Potencier
+Copyright (c) 2022-2022 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/LICENSE]: Please revert

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php (hunk 4 above)
Comment (gnito-org) [reply src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: The priority parameter will be omitted from the API call if null.

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthOptions.php [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    public function getTo(): ?array
+    {
+        return $this->options['to'] ?? null;
+    }
+
+    public function setAccountId(string $accountId): self
+    {
+        $this->options['account_id'] = $accountId;
+        return $this;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthOptions.php:77]:
```suggestion

        return $this;
```

Same for the other methods and PR's 😄

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+        unset($options['application_id']);
+        if (!isset($options['priority']) && $this->priority) {
+            $options['priority'] = $this->priority;
+        }
+        if (!preg_match('/^\+[1-9]\d{1,14}$/', $this->from)) {
+            throw new InvalidArgumentException(sprintf('The "From" number "%s" is not a valid phone number. The number must be in E.164 format.', $this->from));
+        }
+        if (!preg_match('/^\+[1-9]\d{1,14}$/', $message->getPhone())) {
+            throw new InvalidArgumentException(sprintf('The "To" number "%s" is not a valid phone number. The number must be in E.164 format.', $message->getPhone()));
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]:
```suggestion
        
        }
        $opts = $message->getOptions();
        $options = $opts ? $opts->toArray() : [];
        $options['text'] = $message->getSubject();
        $options['from'] = $options['from'] ?? $this->from;
        $options['to'] = $options['to'] ?? [$message->getPhone()];
        $options['account_id'] = $options['account_id'] ?? $this->accountId;
        $options['applicationId'] = $options['application_id'] ?? $this->applicationId;
        unset($options['application_id']);

        if (!isset($options['priority']) && $this->priority) {
            $options['priority'] = $this->priority;
        }

        if (!preg_match('/^\+[1-9]\d{1,14}$/', $this->from)) {
            throw new InvalidArgumentException(sprintf('The "From" number "%s" is not a valid phone number. The number must be in E.164 format.', $this->from));
        }

        if (!preg_match('/^\+[1-9]\d{1,14}$/', $message->getPhone())) {
            throw new InvalidArgumentException(sprintf('The "To" number "%s" is not a valid phone number. The number must be in E.164 format.', $message->getPhone()));
        }
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        $endpoint = sprintf('https://%s/api/v2/users/%s/messages', $this->getEndpoint(), $options['account_id']);
+        unset($options['accountId']);
+        $response = $this->client->request(
+            'POST',
+            $endpoint,
+            [
+                'auth_basic' => $this->username . ':' . $this->password,
+                'json'       => array_filter($options),
+            ]
+        );
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]:
```suggestion

        $response = $this->client->request('POST', $endpoint, [
            'auth_basic' => $this->username . ':' . $this->password,
            'json'       => array_filter($options),
        ]);
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+        try {
+            $statusCode = $response->getStatusCode();
+        } catch (TransportExceptionInterface $e) {
+            throw new TransportException('Could not reach the remote Bandwidth server.', $response, 0, $e);
+        }
+        if (202 !== $statusCode) {
+            $error = $response->toArray(false);
+            throw new TransportException(sprintf('Unable to send the SMS - %s - %s', $error['type'], $error['description']), $response);
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]:
```suggestion

        try {
            $statusCode = $response->getStatusCode();
        } catch (TransportExceptionInterface $e) {
            throw new TransportException('Could not reach the remote Bandwidth server.', $response, 0, $e);
        }

        if (202 !== $statusCode) {
            $error = $response->toArray(false);
            throw new TransportException(sprintf('Unable to send the SMS - %s - %s', $error['type'], $error['description']), $response);
        }

```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+        $success = $response->toArray(false);
+        $sentMessage = new SentMessage($message, (string) $this);
+        $sentMessage->setMessageId($success['id']);
+        return $sentMessage;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php:106]:
```suggestion

        return $sentMessage;
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransportFactory.php [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        $accountId = $dsn->getRequiredOption('account_id');
+        $applicationId = $dsn->getRequiredOption('application_id');
+        $priority = $dsn->getOption('priority');
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $port = $dsn->getPort();
+
+        return (new BandwidthTransport($username, $password, $from, $accountId, $applicationId, $priority, $this->client, $this->dispatcher))
+            ->setHost($host)
+            ->setPort($port)
+        ;
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransportFactory.php]: One line please

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransportFactory.php [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+final class BandwidthTransportFactory extends AbstractTransportFactory
+{
+    private const TRANSPORT_SCHEME = 'bandwidth';
+
+    public function create(Dsn $dsn): BandwidthTransport
+    {
+        $scheme = $dsn->getScheme();
+        if (self::TRANSPORT_SCHEME !== $scheme) {
+            throw new UnsupportedSchemeException($dsn, self::TRANSPORT_SCHEME, $this->getSupportedSchemes());
+        }
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransportFactory.php]:
```suggestion
        $scheme = $dsn->getScheme();
        
        if (self::TRANSPORT_SCHEME !== $scheme) {
            throw new UnsupportedSchemeException($dsn, self::TRANSPORT_SCHEME, $this->getSupportedSchemes());
        }
        
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/LICENSE [hunk 17]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2019-2022 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/LICENSE]:
```suggestion
Copyright (c) 2022 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+use Symfony\Component\Notifier\Test\TransportTestCase;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+final class BandwidthTransportTest extends TransportTestCase
+{
+    public function createTransport(
+        HttpClientInterface $client = null,
+        string $from = 'from'
+    ): BandwidthTransport
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php]: One line please

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    {
+        return new BandwidthTransport('username', 'password', $from, 'account_id', 'application_id', 'priority', $client ?? $this->createMock(HttpClientInterface::class));
+    }
+
+    public function invalidFromProvider(): iterable
+    {
+        // phone numbers
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php]: Not needed

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    /**
+     * @dataProvider invalidFromProvider
+     */
+    public function testInvalidArgumentExceptionIsThrownIfFromIsInvalid(string $from): void
+    {
+        $transport = $this->createTransport(null, $from);
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage(sprintf('The "From" number "%s" is not a valid phone number. The number must be in E.164 format.', $from));
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php:50]:
```suggestion

        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage(sprintf('The "From" number "%s" is not a valid phone number. The number must be in E.164 format.', $from));
        
```

File: src/Symfony/Component/Notifier/LICENSE (hunk 9 above)
Comment (OskarStark) [inline src/Symfony/Component/Notifier/LICENSE]:
```suggestion
Copyright (c) 2019-2022 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        HttpClientInterface $client = null,
+        EventDispatcherInterface $dispatcher = null
+    )
+    {
+        parent::__construct($client, $dispatcher);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('bandwidth://%s?from=%s&account_id=%s&application_id=%s&priority=%s', $this->getEndpoint(), $this->from, $this->accountId, $this->applicationId, $this->priority);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]: If priority is null, it shouldn't be included, should it?

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    public function missingRequiredOptionProvider(): iterable
+    {
+        yield 'missing option: from' => ['bandwidth://username:password@default?account_id=account_id&application_id=application_id&priority=priority'];
+        yield 'missing option: account_id' => ['bandwidth://username:password@default?from=0611223344&application_id=application_id&priority=priority'];
+        yield 'missing option: application_id' => ['bandwidth://username:password@default?from=0611223344&account_id=account_id&priority=priority'];
+    }
+
+    public function supportsProvider(): iterable
+    {
+        yield [true, 'bandwidth://username:password@default?from=0611223344&account_id=account_id&application_id=application_id&priority=priority'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php:44]: Can you please add a test case without priority, thanks

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+final class BandwidthTransportTest extends TransportTestCase
+{
+    public function createTransport(HttpClientInterface $client = null, string $from = 'from'): BandwidthTransport
+    {
+        return new BandwidthTransport('username', 'password', $from, 'account_id', 'application_id', 'priority', $client ?? $this->createMock(HttpClientInterface::class));
+    }
+
+    public function invalidFromProvider(): iterable
+    {
+        // phone numbers
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php]:
```suggestion
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/composer.json [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+            "name": "Symfony Community",
+            "homepage": "https://symfony.com/contributors"
+        }
+    ],
+    "require": {
+        "php": ">=8.1",
+        "symfony/event-dispatcher": "^5.4|^6.0",
+        "symfony/http-client": "^5.4|^6.0",
+        "symfony/notifier": "^6.3"
+    },
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/composer.json:25]: please add `require-dev` section with event-dispatcher: 5.0|6.0

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        yield [true, 'bandwidth://username:password@default?from=0611223344&account_id=account_id&application_id=application_id'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php]: Sorry, this test needs to be moved to `createProvider`

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php (hunk 23 above)
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportTest.php]: Please remove this comment

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+final class BandwidthTransportFactoryTest extends TransportFactoryTestCase
+{
+    public function createFactory(): BandwidthTransportFactory
+    {
+        return new BandwidthTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield ['bandwidth://host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id'];
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php]:
```suggestion
        yield ['bandwidth://host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority'];
        yield ['bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id']
```

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+            throw new InvalidArgumentException(sprintf('The "From" number "%s" is not a valid phone number. The number must be in E.164 format.', $this->from));
+        }
+
+        if (!preg_match('/^\+[1-9]\d{1,14}$/', $message->getPhone())) {
+            throw new InvalidArgumentException(sprintf('The "To" number "%s" is not a valid phone number. The number must be in E.164 format.', $message->getPhone()));
+        }
+        $endpoint = sprintf('https://%s/api/v2/users/%s/messages', $this->getEndpoint(), $options['account_id']);
+        unset($options['accountId']);
+
+        $response = $this->client->request('POST', $endpoint, ['auth_basic' => $this->username . ':' . $this->password, 'json' => array_filter($options)]);
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php]:
```suggestion
        $response = $this->client->request('POST', $endpoint, [
            'auth_basic' => $this->username . ':' . $this->password,
            'json' => array_filter($options),
        ]);
```

Lets make this line more readable

**Comment by OskarStark** [comment]
One minor and a test change, afterwards ready to merge from my side 👍

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php (hunk 27 above)
Comment (gnito-org) [reply src/Symfony/Component/Notifier/Bridge/Bandwidth/BandwidthTransport.php:87]: :face_exhaling: :face_with_head_bandage:

This has been an extremely painful exercise. Not conducive to enthusiastic contribution.

RUN AUTOMATED CODE FORMATTING DURING MERGE OR SOMETHING LIKE THAT, if you guys want to enforce a specific code formatting standard. I'm all for it, but don't burden the contributors with it.

File: src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+{
+    public function createFactory(): BandwidthTransportFactory
+    {
+        return new BandwidthTransportFactory();
+    }
+
+    public function createProvider(): iterable
+    {
+        yield ['bandwidth://host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id&priority=priority'];
+        yield ['bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id']
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Bandwidth/Tests/BandwidthTransportFactoryTest.php]:
```suggestion
        yield ['bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id', 'bandwidth://username:password@host.test?from=0611223344&account_id=account_id&application_id=application_id'];
```

Syntax error

**Comment by gnito-org** [comment]
:face_exhaling: :face_with_head_bandage:

This has been an extremely painful exercise. Not conducive to enthusiastic contribution.

RUN AUTOMATED CODE FORMATTING DURING MERGE OR SOMETHING LIKE THAT, if you guys want to enforce a specific code formatting standard. I'm all for it, but don't burden the contributors with it.

**Comment by OskarStark** [comment]
> This has been an extremely painful exercise. Not conducive to enthusiastic contribution.

I am very sorry, we don't have a very hard coding standard, the reason is, that sometimes we allow it. In this case it's more about "keep all the bridges nearly the same", it makes it much easier to maintain. Most of the stuff was copied from another bridge, which is totally fine, but the "code style" was not copied.

In your case it was even harder, because you opened four PRs in parallel 🤷‍♂️

One test is still failing, maybe you can have a look. Thanks

**Review by OskarStark — APPROVED** [review]

**Review by OskarStark — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @gnito-org.

**Pull request merged**